### DISCUSSION TITLE
Documentation reorg

### DISCUSSION THREAD

**Discussion by 1st1** [geldata/gel#2823] [Ideas]
New proposed navigation structure:

* `Introduction`
  * Changelog 

* Guides
  * `Quickstart`
  * `Cheetsheets`-- The goal here is to actually drop this section, salvaging good content and transforming it into guides. For now let's keep the section, just nest it into the new top-level category `Guides` 
    * `REPL` -> `Interactive Shell`
    * `Admin Commands` -> `Administration` (move down)
    * `CLI` -> `Command Line` (move down)
    * `Select` -> `Selecting Data`
    * `Insert` -> `Inserting Data`
    * `Update` -> `Updating Data`
    * `Delete` -> `Deleting Data`
    * `Functions` -> `User Functions`
    * `Boolean Expressions` (stays the same)
    * `Types` -> `Object Types`
    * `Aliases` -> `Expression Aliases` -- at the top of the page explain briefly what Expression Aliases are and link to `Concepts / Expression Aliases`
    * `Annotations` -> `Schema Annotations`
    * `Migrations`: drop it or rewrite it; right now it's pointless.
    * `Special Syntax` -> `Link Properties` (drop the rest, add more content for link properties, such as insert/update/delete)
  * Installation -> moved from `Administration / Installation`
  * Initial Setup -> moved from `Administration / Initial Setup`
  * Server Setup  -> moved from `Administration / Server Setup`

* `Data Model` -- we move all type and constraints to the `Standard Library`. This section becomes a guide-like introduction to the EdgeDB data model
  * `Overview`: Should talk about the top-level hierarchy first: a running EdgeDB server is called "Instance". "Instances" can have many "Databases" in them. "Databases" have modules in them. Every database has a schema that fully describes its inner structure. "Modules" contain user-defined object and scalar types, functions, etc. Then the current content of the section will follow. *We need to draw an illustration for this section.*
  * `Object types` -> break into sub-categories:
    * `Overview` -> the current first three paragraphs. *Need to have an illustration for this section.*
    * `Base Objects` -> overview of `std::BaseObject` & `std::Object`
    * `User-defined Objects` -> Say that EdgeDB is all about user-defined object types. Show an SDL with a `Movie` type. Explain in English what it means - a container for data. Add a bullet list below linking to: `links`, `properties`, `SDL`, etc.
    * `Free Objects` -> This section should mention `std::FreeObject`, show a very simple `SELECT` query and redirect for more info to `EdgeQL / Selecting Data / Free Shapes`.
  * `Scalar Types` -> The current content, which is basically a nav-tree, must go. It just replicates the navigation on the left. It should be replaced with text explaining how scalar types are different from object types and prompt the user to explore the navigation on the left. *Perhaps we should add an illustration to this section with the tower of scalar types*. All subsections stay as is. Move the types definition to `Standard Library`
  * `Collection Types` -> stays as is
  * `Abstract Types` ->  Move the entire section to `Standard Library`
  * `Functions` -> Add a subsection:
    * `User-defined functions` - show an example.
  * `Links` - stays as is
  * `Properties` -> Split into `Proprties` and a new top-level category `Link Properties`
  * `Computed Links & Properties` -> stays as is
  * `Indexes` -> stays as is
  * `Constraints` -> stays as is
  * `Expression Aliases` -> stays as is
  * `Annotations` -> stays as is

* `EdgeQL`
  * The section-page currently is a decent overview. Keep it, but add links to it.
  * `Overview` -> `Concepts`;  keep the current overview up to [Queries](https://www.edgedb.com/docs/edgeql/overview#queries). The `Queries` section will go into a new sub-category `EdgeQL / Expressions / Evaluation Algorithm`.
  * `Expressions` -> Leave as is
    * `Paths`
    * `Shapes `
    * `Arrays`
    * `Tuples`
    * `Parameters`
    * `Types Language`
    * `Evaluation Algorithm`
  * `Statements` -> `Commands`
  * `Functions and Operators` -> Move to the new top-level `Standard Library`
  * `SDL` stays the same
  * `DDL` stays the same
  * `Introspection` stays the same
  * `Administration` -> stays as is
  * `Lexical Structure` -> stays as is

* `GraphQL`
  * Move content from Cheetsheets
  * Move content from `Clients / GraphQL`

* `Standard Library` -> new section, mainly created out of `EdgeQL / Functions & Operators`
  * Add constraints from `Data Model / Constraints`
  * Add scalars from `Data Model / Scalar Types`
  * Add abstract types from `Data Model / Abstract Types`. 

* `Clients` -> `Client Libraries`. The overview section should actually talk a little bit about our vision: "EdgeDB is a unified data layer, that looks the same and talks the same no matter from language you're interfacing with it." We may add a diagram to this page.
  * `EdgeDB Python Driver` -> `Python`. Ditto for other subcategories.

* `CLI` -> `Command Line`

* `Administration` -> move into `Guides`

* `Internals` stays the same

* `Glossary` - a new section; let's use sphinx glossary functionality.

**Comment by 1st1** [comment]
* Perhaps GraphQL should be elevated into its own top-level directory?

**Reply by vpetrovykh** [reply]
Yes, we need a top-level entry for GraphQL. It's a very recognizable and even familiar technology for a lot of people, for some it might even be what got them interested in EdgeDB in the first place.

**Reply by clmnin** [reply]
Yes, many devs find the GraphQL feature interesting.

Real conversation
```
Me: You should check out EdgeDB. edgedb.com

Them: Hm.. a new database? I don't know. Oh, it has GraphQL out of the box? Looks pretty cool!
<goes down the rabbit hole>
```

**Reply by 1st1** [reply]
Yeah :) Let's elevate it a little.

**Comment by vpetrovykh** [comment]
I would keep "Expression Aliases" as the term because our aliases don't have to be object-based, much like the aliases in the `WITH` block.

**Reply by 1st1** [reply]
OK, let's keep them as is. I'll update the top post

**Comment by vpetrovykh** [comment]
Overall this seems like a good reorg for our docs that are mostly reference docs with one "how-to" section. Once we clean up and nail down the reference docs here we can work on improving the "how-to" and "tutorial" bits. Presumably, the reference will be relatively stable and unchanging (as the core concepts should be stable) so that we can shift our effort into making "how-to" and "tutorial" sections cover more stuff.

**Comment by elprans** [comment]
A good cleanup.

**Comment by 1st1** [comment]
@vpetrovykh OK, you can start implementing this now :)

**Comment by 1st1** [comment]
@vpetrovykh Please do this in a series of relatively small PRs, not as one big one.

**Reply by vpetrovykh** [reply]
OK, will do.

**Comment by colinhacks** [comment]
Just getting around to this now, sorry @vpetrovykh 😣

Everything looks really great. I had one huge recommendation — perhaps we should split Data Model into two top level sections - "Type System" and "Data Modeling"?

Something like this:
```
- Data modeling
  - Scalar types (brief summary of scalar types, a la "Data Types" for Postgres)
  - Properties
  - Links
  - Computed fields (I think this is clear...?)
  - Constraints
  - Indexes
  - Custom scalars (new page)
  - Functions (pull from EdgeQL > SDL > Functions)
  - Expression aliases
  - Annotations
  - Modules (included because it requires the user to learn to appropriate SDL syntax)

- Type System (flatten "Scalar Types")
  - String
  - Numbers
  - Boolean
  - Date and Time
  - JSON
  - Enum
  - Bytes
  - UUID
  - Arrays
  - Tuples (flatten "Collection Types")
  - Object Types (containing the "Overview" and "Base objects" content from Yury's proposal)
  - Casts (flatten "Collection Types")

```
A few points in favor of this:
- Roughly, this splits the docs into built-in stuff and user-defined stuff. 
- It's flatter. Even if we don't split up Type System and Data Modeling, I think it's worth flattening Scalar Types and Collection Types. It's extremely satisfying for a new user to click "Type System" and seeing everything they expect in one place. 
- This split draws attention to the type system, which is one of EdgeDB's strengths/selling points.
- I think a typical user will scan for a listing of scalar types (the analog of Postgres's Data Types page) then want to jump straight into defining a schema. Currently, the structure of the docs makes it feel like you need to read 15+ pages of somewhat arcane docs about the typesystem before you get to the "Links" page. 
- It's weird that things like "Properties" and "Indexes" are in the same navigation menu as "Scalar Types".

I think "Type system" should go after Data Modelling and EdgeQL. The actual content of the Scalar Types & Collection Types pages isn't necessary for most users...they just need a list of type names/syntax for primitive types which should ideally be centralized in a single page. The Type System section is akin to an API Reference than "actual" docs.
  
Other small nits:
- I don't think it makes sense to ever have a page called "Overview"...when you click on a menu item with children, it should show you the "overview" automatically. Having a separate child element called "Overview" is weird. So for instance, we should merge `edgeql/index` and `edgeql/overview`.
- I think there should be a cheatsheet for scalar + collection data types. The analog of "[Data Types](https://www.postgresql.org/docs/9.5/datatype.html)" in Postgres. Perhaps a table so users can mentally map between EdgeDB types and Postgres types.
- I think leaving in Abstract Types makes sense and a single source of reference for all abstract types. Mentioning the relevant abstract types inside the appropriate section (Tuples, etc) also makes sense. Having both is nice...redundancy isn't a bad thing. It lets the reader make connections to things they previously read. 
- The "Deployment" section contains almost no information about deployment. All the content in this section currently uses the CLI which isn't intended for production use (..right?). But this section will shortly be fleshed out (Docker instructions, environment variables, etc) so we can still change it now.

**Reply by 1st1** [reply]
> Everything looks really great. I had one huge recommendation — perhaps we should split Data Model into two top level sections - "Type System" and "Data Modeling"?

> This split draws attention to the type system, which is one of EdgeDB's strengths/selling points.

I'm not sure about this. The term "type system" actually encompasses both object types and scalar types and how they compose/interact with each other. Having just scalar types in a category Type System sounds very confusing to me.

That said, I like the idea of "Data Modeling" category. But IMO it should actually be a group of articles discussing actual schema modelling in EdgeDB - best practices, common approaches, patterns.

> I don't think it makes sense to ever have a page called "Overview"...when you click on a menu item with children, it should show you the "overview" automatically. Having a separate child element called "Overview" is weird. So for instance, we should merge edgeql/index and edgeql/overview.

+1

> I think there should be a cheatsheet for scalar + collection data types. The analog of "Data Types" in Postgres. Perhaps a table so users can mentally map between EdgeDB types and Postgres types.

+1 for adding a howto/cheatsheet

> I think leaving in Abstract Types makes sense and a single source of reference for all abstract types. Mentioning the relevant abstract types inside the appropriate section (Tuples, etc) also makes sense. Having both is nice...redundancy isn't a bad thing. It lets the reader make connections to things they previously read.

Disagree here, because abstract types have only meaning in context of polymorphic functions. You will never see them elsewhere, they are pretty much a distraction. I'm OK with referencing them from other sections. Moreover, in the Scalar Types overview we should have a tree of scalar types, where we can include the abstract types to give them more visibility.

> The "Deployment" section contains almost no information about deployment. All the content in this section currently uses the CLI which isn't intended for production use (..right?). But this section will shortly be fleshed out (Docker instructions, environment variables, etc) so we can still change it now.

True, we (@elprans in particular) will be updating this category in the coming weeks with more content.

**Reply by vpetrovykh** [reply]
Regarding "Type System" and "Data Modeling" split I think there's some good in trying to split "data types" from "data layout", but I'm not sure about the exact details. Namely, the object types (which are mentioned in "Type System") are both the data units AND data layout (with all the links and constraints and such). So the split is less clear here.

**Reply by 1st1** [reply]
What I don't like about the current documentation is that we already have some tightly related information spread between two sections - Data Model and EdgeQL. For example Indexes - Data Model talks about them in an abstract sense, but you really need to read the EdgeQL category to understand what you need to actually use them. @colinhacks proposes to add yet another top-level category, so now instead of looking in two places I'd have to look in 3 (check if Indexes are mentioned in the Type System or Data Modelling, and then read EdgeQL).

I understand the intent of this but I still don't quite like the suggested split. Although I'd be OK with renaming `Concepts` to `Data Modelling`, I think.

**Comment by 1st1** [comment]
* Introduction
  * Changelog

* Guides
  * Quickstart
  * Cheatsheets {should be kept to bare minimum; most should become guides}
  * Data Modeling {Object Types, Links, Properties, Expression Aliases...}
  * Querying Data (Selecting, Grouping, Window Functions later, WITH, etc etc)
  * Mutating Data (INSERT, DELETE, UPDATE}
  * Administration (-> Deployment)
  * Developing EdgeDB (moved from Internals)

* EdgeQL
  * Overview -> Concepts 
  * 

* GraphQL

* Standard Library
  * Scalar Types 
  * Constraints
  * Functions...

* CLI

* Client Libraries

* Internals

* Glossary


1. (Administration -> Deployment) --- let's move it under "Guides"
2. Standard Library should contain Constraints & Scalar Types in it
3. Data Model - to become more howto/reference-level 
4. Changelog becomes a sub-category of Introduction

**Reply by 1st1** [reply]
Consider adding a separator for the navitems saying "reference"

**Comment by 1st1** [comment]
@colinhacks @vpetrovykh @elprans I've updated the top-level post. I decided that we should keep the Data Model section (we can consider renaming it). It is a guide-like introduction to the EdgeDB data model and I think it's important for it to continue being visible, as this is literally *the* resource to see all the concepts of EdgeDB at one place. Other than that it reflects what we discussed.

**Comment by vpetrovykh** [comment]
After a discussion on the Friday call we've come to the conclusion that there needs to be an "Operators" subsection in "EdgeQL" that should at least list the available operators, maybe introduce a few examples and then link to the definitions in "Standard Library".

**Reply by 1st1** [reply]
> After a discussion on the Friday call we've come to the conclusion that there needs to be an "Operators" subsection in "EdgeQL" that should at least list the available operators,

Listing them is fine, but the majority of the documentation must be in the Standard Library section.
